py3status/modules/emerge_status.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+
+"""
+Emerge_status provides a short information on the current
+emerge status if there is an emerge process currently running
+
+Configuration parameters:
+    cache_timeout: how often we refresh this module in second (default 1)
+    emerge_log_file: in case it is placed somewhere else than the default
+        (default '/var/log/emerge.log')
+    format: Display format to user (default '{current} / {total}')
+    hide_if_zero: Don't show in bar if there is no emerge running (default True)
+
+    More format options:
+        Ex: '{current} / {total}'
+        Result: "3 / 14"
+        Ex: '{current} / {total} = {'
+        Result: "3 / 14 = py3status"
+        Ex: '{category} - {pkg}'
+        Result: "x11-misc - py3status"
+
+@author: AnwariasEu
+"""
+
+import os
+import re
+
+
+class Py3status:
+    """
+    """
+    cache_timeout = 1
+    emerge_log_file = '/var/log/emerge.log'
+    format = u'{current} / {total}'
+    hide_if_zero = True
+
+    def _emergeRunning(self):
+        """
+        Use pgrep to check if emerge is running.
+
+        Returns true if at least one instance of emerge is running.
+        """
+        tmp = os.popen("pgrep emerge").read()
+        pcount = tmp.count("\n")
+        if pcount > 0:
+            return True
+        else:
+            return False
+
+    def _getProgress(self):
+        """
+        Get current progress of emerge.
+
+        returns a dict containing current and total value.
+        """
+        ret = {}
+        ret['current'] = None
+        ret['total'] = None
+        ret['err'] = False
+        ret['err_text'] = ""
+        ret['category'] = None
+        ret['pkg'] = None
+
+        input_data = []
+
+        """
+        Try to open file, if unable throw error to be catched later
+        """
+        try:
+            with open(self.emerge_log_file, 'r') as fp:
+                for line in fp:
+                    input_data.append(line)
+        except IOError as err:
+            ret['err'] = True
```

Comment:
This could be better off with `self.py3.error()` instead. I don't know how often this can happen so if it's not something very often, then we could skip `Try/Except` instead. The exception would be shown on the bar.

We also could check for binary `emerge` in `post_config_hook` so this can't happen on other distributions.

Reply:
This can happen if the file does not exist e.g. the module is used on system where there is no portage and therefore no emerge.log.
But it can also happen the user has no read access to the file, which happens in case the user is not part of the 'portage' group.

Error handling can definitely be improved.

Reply:
We could make a `STRING_NOT_INSTALLED` (see `moc.py`) so this would only run on machines with installed `emerge` == portage (?). I wonder if people who uses `emerge` regularly may already add themselves to `portage` group? If not, we could make a small note about `portage` group. Then we might not have to worry about error handling anymore. And if there was one, users can read the exception on the bar instead of custom error string.

Reply:
emerge is the main binary of the portage package manager.
Checking for emerge binary implemented in 42e811f.
Error handling addressed in  e9d9ab8.